http-clients/netty-nio-client/src/main/java/software/amazon/awssdk/http/nio/netty/internal/http2/Http2ToHttpInboundAdapter.java

```diff
@@ -56,6 +63,30 @@ protected void channelRead0(ChannelHandlerContext ctx, Http2Frame frame) throws
         }
     }
 
+    @Override
+    public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) {
+        if ((cause instanceof ReadTimeoutException || cause instanceof WriteTimeoutException) && ctx.channel().parent() != null) {
+            healthCheck(ctx);
+        }
+
+        ctx.fireExceptionCaught(cause);
+    }
+
+    private void healthCheck(ChannelHandlerContext ctx) {
```

Comment:
This is just a thought, but it might make it cleaner to do the health checking and management of the keep alive flag as a separate handler that gets added/removed from the parent pipeline as necessary

Reply:
@dagnir Are you saying moving the keep alive flag management into a separate parent handler?

The exception here happens on a child channel and we have to set `KEEP_ALIVE  = false` for the parent channel in this child channel handler.

I can move this healthcheck logic into a separate handler, say `Http2FrameExceptionHandler`.

Reply:
Yes, basically what I had in mind was that if we determine that we need to do a health check, then instead of taking care of that in this handler, add a new handler to the pipeline before this one that takes care of the health checking like constructing/writing the frame, setting the flag, checking the ACK etc.

Reply:
I see. But we only know we need to do health check when there's a predefined exception thrown from a child channel and even though we create a parent handler for that, I'm not sure how we can trigger the parent handler to do the health check from the child handler.

I guess we can create a specific UserEvent to trigger the parent channel to do health check, but I feel like that's overly complicated.

Reply:
Discussed offline. To sum up, we can add a parent handler to handle all PING frames when this happens, something like  `ctx.channel().parent().pipeline().addLast(Http2PingFrameHandler.create()); ` 
and use `handlerAdded` callback to send PING frame.

There is a concern that adding a handler can add delays and we need to run perf tests to be sure. For now, we are just going to send the PING frame from a separate child handler.